sinker/generate_csv_sinker.go

```diff
@@ -0,0 +1,311 @@
+package sinker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/streamingfast/dstore"
+	"github.com/streamingfast/logging"
+	"github.com/streamingfast/shutter"
+	sink "github.com/streamingfast/substreams-sink"
+	pbdatabase "github.com/streamingfast/substreams-sink-database-changes/pb/sf/substreams/sink/database/v1"
+	"github.com/streamingfast/substreams-sink-postgres/bundler"
+	"github.com/streamingfast/substreams-sink-postgres/bundler/writer"
+	"github.com/streamingfast/substreams-sink-postgres/db"
+	"github.com/streamingfast/substreams-sink-postgres/state"
+	pbsubstreamsrpc "github.com/streamingfast/substreams/pb/sf/substreams/rpc/v2"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+type GenerateCSVSinker struct {
+	*shutter.Shutter
+	*sink.Sinker
+	destFolder string
+
+	fileBundlers map[string]*bundler.Bundler
+	stopBlock    uint64
+
+	// cursor
+	stateStore state.Store
+	bundleSize uint64
+
+	loader *db.Loader
+	logger *zap.Logger
+	tracer logging.Tracer
+
+	stats *Stats
+}
+
+func NewGenerateCSVSinker(
+	sink *sink.Sinker,
+	destFolder string,
+	workingDir string,
+	bundleSize uint64,
+	bufferSize uint64,
+	loader *db.Loader,
+	logger *zap.Logger,
+	tracer logging.Tracer,
+) (*GenerateCSVSinker, error) {
+	blockRange := sink.BlockRange()
+	if blockRange == nil || blockRange.EndBlock() == nil {
+		return nil, fmt.Errorf("sink must have a stop block defined")
+	}
+
+	// create cursor in working Dir
+	stateStorePath := filepath.Join(workingDir, "state.yaml")
+	stateFileDirectory := filepath.Dir(stateStorePath)
+	if err := os.MkdirAll(stateFileDirectory, os.ModePerm); err != nil {
+		return nil, fmt.Errorf("create state file directories: %w", err)
+	}
+
+	stateDStore, err := dstore.NewStore(destFolder, "", "", false)
+	if err != nil {
+		return nil, err
+	}
+	stateStore, err := state.NewFileStateStore(stateStorePath, stateDStore, logger)
+	if err != nil {
+		return nil, fmt.Errorf("new file state store: %w", err)
+	}
+
+	s := &GenerateCSVSinker{
+		Shutter: shutter.New(),
+		Sinker:  sink,
+
+		fileBundlers: make(map[string]*bundler.Bundler),
+		stopBlock:    *blockRange.EndBlock(),
+
+		loader: loader,
+		logger: logger,
+		tracer: tracer,
+
+		stateStore: stateStore,
+		bundleSize: bundleSize,
+
+		stats: NewStats(logger),
+	}
+
+	csvOutputStore, err := dstore.NewStore(destFolder, "csv", "", false)
+	if err != nil {
+		return nil, err
+	}
+	tables := s.loader.GetAvailableTablesInSchemaList()
+
+	for _, table := range tables {
+		columns := s.loader.GetColumnsForTable(table)
+		fb, err := getBundler(table, s.Sinker.BlockRange().StartBlock(), s.stopBlock, bundleSize, bufferSize, csvOutputStore, workingDir, logger, columns)
+		if err != nil {
+			return nil, err
+		}
+		s.fileBundlers[table] = fb
+	}
+
+	return s, nil
+}
+
+func (s *GenerateCSVSinker) Run(ctx context.Context) {
+	s.stateStore.Start(ctx)
```

Comment:
The `stateStore`' shutter is not connected back to its "parent". If an error occurs while the state store is running, it will not be reported anywhere.

```suggestion
    s.stateStore.OnTerminating(s.Shutdown)
	s.stateStore.Start(ctx)
```

Reply:
Ok, done